romancal/ramp_fitting/ramp_fit_step.py

```diff
@@ -7,76 +7,179 @@
+
+        Returns
+        -------
+        out_model : ImageModel
+            Model containing a count-rate image.
+        """
+        max_cores = self.maximum_cores
+        input_model.data = input_model.data[np.newaxis, :]
+        input_model.groupdq = input_model.groupdq[np.newaxis, :]
+        input_model.err = input_model.err[np.newaxis, :]
+
+        log.info(f"Using algorithm = {self.algorithm}")
+        log.info(f"Using weighting = {self.weighting}")
+
+        buffsize = ramp_fit.BUFSIZE
+        image_info, integ_info, opt_info, gls_opt_model = ramp_fit.ramp_fit(
+            input_model,
+            buffsize,
+            self.save_opt,
+            readnoise_model.data.value,
+            gain_model.data.value,
+            self.algorithm,
+            self.weighting,
+            max_cores,
+            dqflags.pixel,
+        )
+        readnoise_model.close()
+        gain_model.close()
 
+        # Save the OLS optional fit product, if it exists
+        if opt_info is not None:
+            opt_model = create_optional_results_model(input_model, opt_info)
+            self.save_model(opt_model, "fitopt", output_file=self.opt_name)
+
+        # All pixels saturated, therefore returning an image file with zero data
+        if image_info is None:
+            log.info("All pixels are saturated. Returning a zeroed-out image.")
+
+            # Image info order is: data, dq, var_poisson, var_rnoise, err
+            image_info = (
+                np.zeros(input_model.data.shape[2:], dtype=input_model.data.dtype),
+                input_model.pixeldq
+                | input_model.groupdq[0][0]
+                | dqflags.group["SATURATED"],
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+            )
+
+        out_model = create_image_model(input_model, image_info)
+        return out_model
+
+    def ols_cas22(self, input_model, readnoise_model, gain_model):
+        """Peform Optimal Linear Fitting on arbitrarily space resulants
+
+        Parameters
+        ----------
+        input_model : RampModel
+            Model containing ramps.
+
+        readnoise_model : ReadnoiseRefModel
+            Model with the read noise reference information.
+
+        gain_model : GainRefModel
+            Model with the gain reference information.
+
+        Returns
+        -------
+        out_model : ImageModel
+            Model containing a count-rate image.
+        """
+        resultants = input_model.data.value
+        dq = input_model.groupdq
+        read_noise = readnoise_model.data.value
+        gain = gain_model.data.value
+        read_pattern = input_model.meta.exposure.read_pattern
+
+        # add in the gain
+        resultants_adjusted = resultants * gain
+        read_noise_adjusted = read_noise * gain
```

Comment:
Can we do these in place?  In particular, the resultants_adjusted, which is large?